Distinguishes between TemplateVM & AppVM in log collection

conorsch opened the pull request:
## Status

Ready for review.

## Description of Changes

Fixes #476

Changes proposed in this pull request:

* Consolidates the log-aggregation logic into a single SLS file, to DRY out the config
* Stores logs separately for AppVM and TemplateVM clients in `sd-log:~/QubesIncomingLogs/`
* Adds additional config tests for logging config, and RPC policies

## Testing


These changes require a patch to the `securedrop-log` package, so in order to test, you'll need to build a custom package. Details below.

You'll need to apply the workaround https://github.com/freedomofpress/securedrop-workstation/issues/485#issuecomment-595244676 in order to review effectively.

- [ ] Check out https://github.com/freedomofpress/securedrop-log/pull/16 and create new tarball (`python3 setup.py sdist`)
- [ ] Check out https://github.com/freedomofpress/securedrop-debian-packaging/pull/150 and build new deb pkg (`PKG_VERSION=0.1.1 make securedrop-log`)
- [ ] Copy new 0.1.1 securedrop-log deb package to `sd-workstation/securedrop-log_0.1.1+buster_all.deb`, so you can install it as part of testing without promoting to apt repo
- [ ] `make clone && make clean && make all && make test` passes without error
- [ ] Inspect `sd-log:/home/user/QubesIncomingLogs/` and ensure that both TemplateVM names and AppVM names are present as subdirs
- [ ] **Drop the commit prefixed with "TEMPORARY" prior to merge.** That commit exists solely to facilitate local testing. 

## Checklist

### If you have made code changes

- [x] Linter (`make flake8`) passes in the development environment (this box may
      be left unchecked, as `flake8` also runs in CI)

### If you have made changes to the provisioning logic

- [x] All tests (`make test`) pass in `dom0` of a Qubes install

- [ ] This PR adds/removes files, and includes required updates to the packaging
      logic in `MANIFEST.in` and `rpm-build/SPECS/securedrop-workstation-dom0-config.spec`

^ haven't updated the spec, nor the version; local build passes successfully

conorsch commented:
Rebased on top of latest master (8f60f29dd3c759e1867196128f606775f094556c), including changes from #486.

conorsch commented:
> Inspect `sd-log:/home/user/QubesIncomingLogs/` and ensure that both TemplateVM names and AppVM names are present as subdirs

Come to think of it, that deserves a config test. Added a few service-related checks in this PR, but we should also guard against regressions.

conorsch commented on dom0/sd-workstation.top:
```diff
@@ -38,15 +40,12 @@ base:
     - sd-sys-firewall-files
```
Good idea. I've added a config test for this case. Since a few of the VMs customize the rsyslog config on boot, via rc.local, there may be a race where the log messages you identify can still slip in. Currently, the config test expects zero occurrences. If it turns out one or two slip in, we can update the logic to be <=2 or so.

conorsch commented:
@emkll Great review, please take another look. Given the special case of the whonix VMs, it might be preferable to review #488 first, then drop the `sd-proxy`-related customizations in the logging config here.

emkll commented:
I'm running into some issues when running make clean / make clone / make all / make test, are you seeing these? It might be due to changes in whonix-gw-template that were not reverted properly when i ran make clean. Do you think twe should review https://github.com/freedomofpress/securedrop-workstation/pull/488 first, as the perhaps it may have benefits for the issues i've observed

conorsch commented:
Addressing review comments now, will rebase to include changes from #487, then ping for re-review.

conorsch commented:
Ready for re-review.

Addressed all comments. Opted to disable logging on `sd-gpg` explicitly, simply to match what's already on master, and to focus on the Template/App distinction. Config test suite is passing consistently, across multiple rebuilds, including the newly added tests.

emkll commented:
Thanks @conorsch for the changes, and your patience, this looks good to merge from my perspective.

Functionally tested as follows:
- [x] Check out https://github.com/freedomofpress/securedrop-log/pull/16 and create new tarball (`python3 setup.py sdist`)
- [x] Check out https://github.com/freedomofpress/securedrop-debian-packaging/pull/150 and build new deb pkg (`PKG_VERSION=0.1.1 make securedrop-log`)
- [x] Copy new 0.1.1 securedrop-log deb package to `sd-workstation/securedrop-log_0.1.1+buster_all.deb`, so you can install it as part of testing without promoting to apt repo
- [x] `make clone && make clean && make all && make test` passes without error
- [x] Inspect `sd-log:/home/user/QubesIncomingLogs/` and ensure that both TemplateVM names and AppVM names are present as subdirs

> Opted to disable logging on sd-gpg explicitly

Confirmed, however there is an sd-gpg folder and syslog that contains logs from provisioning-time. I think this is fine, but we should be aware of this

One new change introduced by this PR is that disposable VMs will now have their own folder in QubesIncomingLogs (eg. `disp1234`). This is good: We will be able to more easily debug specific errors without a massive `sd-viewer` log.

@conorsch if you want to drop the final commit, I think we can merge this alongside the securedrop-log and packaging changes. I do think that it might be good to capture this diff somewhere (wiki?) as it was extremely useful in the course of testing this pr.

conorsch commented:
Thanks @emkll & @kushaldas for the detailed review. 

> if you want to drop the final commit, I think we can merge this alongside the securedrop-log and packaging changes.

Done: temporary commit referencing a locally built package has been dropped.

> I do think that it might be good to capture this diff somewhere (wiki?) as it was extremely useful in the course of testing this pr.

Also done, see here: https://github.com/freedomofpress/securedrop-workstation/wiki/Evaluating-new-deb-package-behavior

emkll commented:
securedrop-log and packaging changes have been merged per https://github.com/freedomofpress/securedrop-workstation/pull/487#pullrequestreview-373016915, merging this now
